### ISSUE TITLE
Table returns inconsistent results when rows= argument is a generator

### ISSUE THREAD

**Issue by Cadair** [astropy/astropy#9290]
I tried to construct a Table with rows which come out of a generator (originally a call to `filter`), and for some reason it skips one result when you do this.

```python
In [12]: from astropy.table import Table                                                                                                                                                                                                                                                                       

In [13]: arow = lambda x: [random.random() for i in range(x)]                                                                                                                                                                                                                                                  

In [14]: it = (arow(3) for i in range(10))                                                                                                                                                                                                                                                                     

In [15]: li = [arow(3) for i in range(10)]                                                                                                                                                                                                                                                                     

In [16]: Table(rows=it)           
```
```                                                                                                                                                                                                                                                                             
Out[16]: 
<Table length=9>
        col0                col1                col2       
      float64             float64             float64      
------------------- ------------------- -------------------
...
```
```python
In [17]: Table(rows=li)                                                                                                                                                                                                                                                                                        
Out[17]: 
```
```
<Table length=10>
        col0                 col1                col2       
      float64              float64             float64      
-------------------- ------------------- -------------------
...
```

**Comment by pllim** [comment]
If you do this instead
```python
import numpy as np
it = (np.arange(3) * i for i in range(10))                             
li = [np.arange(3) * i for i in range(10)]                             
```
you will notice that for the case of `it`, the first row is dropped. Maybe this will hint at where the bug is...

**Comment by taldcroft** [comment]
The problem is pretty evident below in the code that handles the `rows` argument.  In the `all(isinstance(...` line it starts going through the generator, and gets through the first one before failing out of `all()`, but then the first item in the generator is gone.
```
        # Row-oriented input, e.g. list of lists or list of tuples, list of
        # dict, Row instance.  Set data to something that the subsequent code
        # will parse correctly.
        is_list_of_dict = False
        if rows is not None:
            if data is not None:
                raise ValueError('Cannot supply both `data` and `rows` values')
            if all(isinstance(row, dict) for row in rows):
                is_list_of_dict = True  # Avoid doing the all(...) test twice.
                data = rows
            elif isinstance(rows, self.Row):
                data = rows
            else:
                data = list(zip(*rows))
```
I think that inserting this (after the `if data is not None` block) would work:
```
if isinstance(rows, types.GeneratorType):
    rows = list(row)
```
Then you have to make a copy of the generator data in memory, but I can't think of any way to get through the `all(..)` test without using up some or all of the generator.

**Comment by taldcroft** [comment]
Fix in #9315.
